pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+		}
+		return false, nil
+	}
+
+	// get the current node price based on the offering
+	// fallback if we can't find the specific zonal pricing data
+	candidatePrice, err := c.getCandidatePrices(ctx, candidates...)
+	if err != nil {
+		return false, fmt.Errorf("getting offering price from candidate node, %w", err)
+	}
+
+	// Check if there are any on-demand nodes, so we can know if we're doing spot to spot consolidation
+	_, allExistingAreSpot := lo.Find(candidates, func(candidate *state.StateNode) bool {
+		return candidate.Labels()[v1beta1.CapacityTypeLabelKey] == v1beta1.CapacityTypeOnDemand
+	})
+
+	// sort the instanceTypes by price before we take any actions like truncation for spot-to-spot consolidation or finding the nodeclaim
+	// that meets the minimum requirement after filteringByPrice
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions = results.NewNodeClaims[0].InstanceTypeOptions.OrderByPrice(results.NewNodeClaims[0].Requirements)
+
+	if allExistingAreSpot &&
+		results.NewNodeClaims[0].Requirements.Get(v1beta1.CapacityTypeLabelKey).Has(v1beta1.CapacityTypeSpot) {
+		return c.validateSpotToSpotConsolidation(ctx, candidates, results, candidatePrice)
+	}
+
+	var incompatibleMinReqKey string
+	// filterByPriceWithMinValues returns the instanceTypes that are lower priced than the current candidate and the requirement for the NodeClaim that does not meet minValues.
+	// If we use this directly for spot-to-spot consolidation, we are bound to get repeated consolidations because the strategy that chooses to launch the spot instance from the list does
+	// it based on availability and price which could result in selection/launch of non-lowest priced instance in the list. So, we would keep repeating this loop till we get to lowest priced instance
+	// causing churns and landing onto lower available spot instance ultimately resulting in higher interruptions.
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions, incompatibleMinReqKey, _ =
+		disruptionutils.FilterByPriceWithMinValues(results.NewNodeClaims[0].InstanceTypeOptions, results.NewNodeClaims[0].Requirements, candidatePrice)
+
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) == 0 {
+		if len(candidates) == 1 {
+			if len(incompatibleMinReqKey) > 0 {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, fmt.Sprintf("minValues requirement is not met for %s", incompatibleMinReqKey))...)
+			} else {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace with a cheaper node")...)
+			}
+		}
+		return false, nil
+	}
+	return true, nil
+}
+
+// propagates the consolidation decision to cluster state and also marks all
+// associated nodeclaims as underutilized
+func (c *Controller) setNodeClaimsUnderutilized(ctx context.Context, setTrue bool, commandID types.UID, candidates ...*state.StateNode) error {
```

Comment:
I think you know this, but I've never been a fan of this false/true syntax. Why not just inline this call?